crypto/rsa/rsa_ossl.c

```diff
@@ -664,10 +668,37 @@ static int rsa_ossl_mod_exp(BIGNUM *r0, const BIGNUM *I, RSA *rsa, BN_CTX *ctx)
     }
 
     if (rsa->flags & RSA_FLAG_CACHE_PUBLIC)
-        if (!BN_MONT_CTX_set_locked
-            (&rsa->_method_mod_n, rsa->lock, rsa->n, ctx))
+        if (!BN_MONT_CTX_set_locked(&rsa->_method_mod_n, rsa->lock,
+                                    rsa->n, ctx))
+            goto err;
+
+    if (smooth) {
+        if (/* m1 = I moq q */
+            !bn_from_mont_fixed_top(m1, I, rsa->_method_mod_q, ctx)
```

Comment:
As with the similar reduction below, rely on the looser bounds for Montgomery reduction, that `m1` is bounded by `R*q`, not `q^2`. `m1` is bounded by `p*q`, so for the smaller prime, this only works because the two primes are the same size. Might be worth a comment to that effect.